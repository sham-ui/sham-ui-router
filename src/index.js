import { DI, inject } from 'sham-ui';
import Navigo from 'navigo';

export default class Router {
    @inject UI = 'sham-ui';

    constructor( root = null, useHash = false, hash = '#', autoResolve = true ) {
        DI.bind( 'router', this );
        this.router = new Navigo( root, useHash, hash );
        this.activePageWidget = null;
        this.activePageOptions = null;
        this.activePageLinks = new Set();
        if ( autoResolve ) {
            this.UI.render.on( 'RegistrationComplete', this.resolve.bind( this ) );
        }
    }

    resolve() {
        this.router.resolve();
    }

    bindPage( url, name, pageWidget, widgetOptions ) {
        this.on( url, {
            as: name,
            uses: () => {
                this.activePageWidget = pageWidget;
                this.activePageOptions = widgetOptions;
                this._renderActivatePage();
            }
        } );
        return this;
    }

    _renderActivatePage() {
        const widget = DI.resolve( 'widgets:active-page-container' );
        if ( undefined !== widget ) {
            this.UI.render.ONLY( widget.ID );
            this.activePageLinks.forEach( x => x.update() )
        }
    }

    on() {
        this.router.on.apply( this.router, arguments );
    }

    off() {
        this.router.off.apply( this.router, arguments );
    }

    notFound() {
        this.router.notFound.apply( this.router, arguments );
    }

    navigate() {
        this.router.navigate.apply( this.router, arguments );
    }

    hooks() {
        this.router.hooks.apply( this.router, arguments );
    }

    destroy() {
        this.router.destroy.apply( this.router, arguments )
    }

    link() {
        return this.router.link.apply( this.router, arguments );
    }

    lastRouteResolved() {
        return this.router.lastRouteResolved.apply( this.router, arguments );
    }

    generate() {
        return this.router.generate.apply( this.router, arguments );
    }

    _registerActivePageLink( widget ) {
        this.activePageLinks.add( widget );
    }

    _unregisterActivePageLink( widget ) {
        this.activePageLinks.delete( widget );
    }
}