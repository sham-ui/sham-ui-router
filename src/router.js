import { DI, inject } from 'sham-ui';
import Navigo from 'navigo';

export default class Router {
    @inject UI = 'sham-ui';

    constructor( root = null, useHash = false, hash = '#' ) {
        DI.bind( 'router', this );
        this.router = new Navigo( root, useHash, hash );
        this.activePageUrl = null;
        this.activePageWidget = null;
        this.activePageOptions = null;
    }

    resolve() {
        this.router.resolve.apply( this.router, arguments );
    }

    bindPage( url, name, pageWidget, widgetOptions ) {
        this.on( url, {
            as: name,
            uses: () => {
                this.activePageUrl = url;
                this.activePageWidget = pageWidget;
                this.activePageOptions = widgetOptions;
                this._renderActivatePage();
            }
        } );
        return this;
    }

    _renderActivatePage() {
        const widget = DI.resolve( 'widgets:router' );
        if ( undefined !== widget ) {
            this.UI.render.ONLY( widget.ID );
        }
    }

    on() {
        this.router.on.apply( this.router, arguments );
    }

    off() {
        this.router.off.apply( this.router, arguments );
    }

    notFound() {
        this.router.notFound.apply( this.router, arguments );
    }

    navigate() {
        this.router.navigate.apply( this.router, arguments );
    }

    hooks() {
        this.router.hooks.apply( this.router, arguments );
    }

    destroy() {
        this.router.destroy.apply( this.router, arguments )
    }

    link() {
        return this.router.link.apply( this.router, arguments );
    }

    lastRouteResolved() {
        return this.router.link.apply( this.router, arguments );
    }

    generate() {
        return this.router.generate.apply( this.router, arguments );
    }
}