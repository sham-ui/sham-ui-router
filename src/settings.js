export const ACTIVE_PAGE_CONTAINER_TYPE = 'active-page-container';
export const ACTIVE_PAGE_LINK_TYPE = 'link-to-active-page';
