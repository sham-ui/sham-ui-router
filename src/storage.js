import createStorage from 'sham-ui-data-storage';
/**
 * Data storage for router service
 * @see https://github.com/sham-ui/sham-ui-data-storage
 * @typedef {Object} RouterStorage
 * @property {string} url Current page url
 * @property {string} name Current page name
 * @property {Object} params Current page params
 * @property {Class<Component>} activePageComponent Current page component class
 * @property {Object} activePageOptions Options for current page component
 */

export const { storage, useStorage } = createStorage( {
    url: '',
    name: '',
    params: {},
    activePageComponent: null,
    activePageOptions: null
}, {
    DI: 'router:storage',
    sync: true
} );
