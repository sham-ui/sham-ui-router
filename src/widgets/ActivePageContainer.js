import { DI } from 'sham-ui';
import ActivePageContainer from './ActivePageContainer.sht';

export default class extends ActivePageContainer {
    constructor() {
        super( ...arguments );
        DI.bind( 'widgets:active-page-container', this );
    }

    render() {
        super.render( ...arguments );
        const router = DI.resolve( 'router' );
        this.remove();
        // eslint-disable-next-line no-undef
        ShamUIView.insert(
            this,
            this.container,
            { ref: null },
            router.activePageWidget,
            router.activePageOptions
        );
    }
}
