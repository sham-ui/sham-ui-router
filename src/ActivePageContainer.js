import { $ } from 'sham-ui-macro/ref.macro';
import { Component, insert } from 'sham-ui';
import { useStorage } from './storage';

/**
 * Options for LinkTo
 * @typedef {Object} LinkToOptions
 * @property {string} path Name of page for link. Default ''
 * @property {Object} params Params of page for link. Default {}
 * @property {string} text Text for link. Default ''
 * @property {boolean} useActiveClass Use activeClass options for active page link. Default false
 * @property {string} activeClass Class name for active link. Default 'active'
 */

/**
 * Component for link to page
 * @class LinkTo
 * @classdesc
 * @property {LinkToOptions} options
 */

/**
 * Component-container for page
 * @example
 * {% import ActivePageContainer from 'sham-ui-router/active-page-container' %}
 * ...
 * <ActivePageContainer/>
 * ...
 */
function ActivePageContainer() {
    this.ref = null;

    let lastRenderedURL = null;
    let lastRenderedComponent = null;

    this.spots = [ [
        $.routerData,
        routerData => {
            const url = routerData.url;
            const activePageComponent = routerData.activePageComponent;
            if (
                url !== lastRenderedURL ||

                // Component can changed without change url
                activePageComponent !== lastRenderedComponent
            ) {

                // Remove self from parent's children map or child ref.
                if ( this.unbind ) {
                    this.unbind();
                }

                // Remove all nested views.
                let i = this.nested.length;
                while ( i-- ) {
                    this.nested[ i ].remove();
                }
            }
            if ( null === activePageComponent ) {
                return;
            }
            insert(
                this,
                this.container,
                this,
                activePageComponent,
                routerData.activePageOptions,
                this.owner,
                this.blocks
            );
            lastRenderedURL = url;
            lastRenderedComponent = activePageComponent;
        }
    ] ];
}

export default Component( useStorage( $.routerData ), ActivePageContainer );
