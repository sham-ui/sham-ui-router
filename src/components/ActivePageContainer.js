import { Component } from 'sham-ui';
import { useStorage } from '../storage';

/**
 * Options for LinkTo
 * @typedef {Object} LinkToOptions
 * @property {string} path Name of page for link. Default ''
 * @property {Object} params Params of page for link. Default {}
 * @property {string} text Text for link. Default ''
 * @property {boolean} useActiveClass Use activeClass options for active page link. Default false
 * @property {string} activeClass Class name for active link. Default 'active'
 */

/**
 * Component for link to page
 * @class LinkTo
 * @classdesc
 * @property {LinkToOptions} options
 */

/**
 * Component-container for page
 * @example
 * {% import ActivePageContainer from 'sham-ui-router/active-page-container' %}
 * ...
 * <ActivePageContainer/>
 * ...
 */
class ActivePageContainer extends Component {
    constructor() {
        super( ...arguments );
        this.lastRendererdURL = null;
        this.lastRenderedComponent = null;
        this.ref = null;
    }

    updateSpots() {
        const url = this.options.routerData.url;
        const activePageComponent = this.options.routerData.activePageComponent;
        if (
            null !== this.lastRendererdURL &&
            (
                url !== this.lastRendererdURL ||

                // Component can changed without change url
                activePageComponent !== this.lastRenderedComponent
            )
        ) {
            this._clearContainer();
        }
        if ( null === activePageComponent ) {
            return;
        }
        window.__UI__.insert(
            this,
            this.container,
            this,
            activePageComponent,
            this.options.routerData.activePageOptions,
            this.owner,
            this.blocks
        );
        this.lastRendererdURL = url;
        this.lastRenderedComponent = activePageComponent;
    }

    _clearContainer() {

        // Remove self from parent's children map or child ref.
        if ( this.unbind ) {
            this.unbind();
        }

        // Remove all nested views.
        let i = this.nested.length;
        while ( i-- ) {
            this.nested[ i ].remove();
        }
    }
}

export default useStorage( 'routerData' )( ActivePageContainer );
