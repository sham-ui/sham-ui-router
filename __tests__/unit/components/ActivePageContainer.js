import { DI } from 'sham-ui';
import { storage } from '../../../src/storage';
import ActivePageContainer from '../../../src/components/ActivePageContainer';
import renderer, { compile } from 'sham-ui-test-helpers';

afterEach( () => {
    DI.resolve( 'router:storage' ).reset();
} );

it( 'renders correctly', () => {
    const generateMock = jest.fn();

    DI.bind( 'router', {
        storage,
        generate: generateMock
    } );
    storage.activePageComponent = compile`
        <h1>Title</h1>
        <div>Content for dummy component</div>
    `;
    storage.activePageOptions = {
        path: 'base',
        text: 'Base page'
    };
    storage.sync();

    generateMock.mockReturnValue( '/base' );

    const meta = renderer( ActivePageContainer, {} );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'correct process dynamic placet', () => {
    const generateMock = jest.fn();

    DI.bind( 'router', {
        storage,
        generate: generateMock
    } );
    storage.activePageComponent = compile`
        <h1>Title</h1>
        <div>Content for dummy component</div>
    `;
    storage.activePageOptions = {
        path: 'base',
        text: 'Base page'
    };
    storage.sync();

    generateMock.mockReturnValue( '/base' );

    const meta = renderer(
        compile( {
            ActivePageContainer
        } )`
        {% if withWrapper %}
            <div class="wrapper">
                <ActivePageContainer/>
            </div>
        {% else %}
            <ActivePageContainer/>
        {% endif %}
        `,
        {
            withWrapper: false
        }
    );
    expect( meta.toJSON() ).toMatchSnapshot();
    meta.component.update( {
        withWrapper: true
    } );
    expect( meta.toJSON() ).toMatchSnapshot();
} );
