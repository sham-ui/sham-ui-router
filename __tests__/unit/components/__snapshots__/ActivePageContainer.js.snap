// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correct process dynamic placet 1`] = `
Object {
  "Constructor": "dummy",
  "Options": Object {
    "ID": "component",
    "withWrapper": false,
  },
  "html": "
<h1>Title</h1>
<div>Content for dummy component</div>
<!--ActivePageContainer-->
<!--if-->
",
}
`;

exports[`correct process dynamic placet 2`] = `
Object {
  "Constructor": "dummy",
  "Options": Object {
    "ID": "component",
    "withWrapper": true,
  },
  "html": "
<div class=\\"wrapper\\">
  <h1>Title</h1>
  <div>Content for dummy component</div>
</div>
<!--if-->
",
}
`;

exports[`renders correctly 1`] = `
Object {
  "Constructor": "ActivePageContainer",
  "Options": Object {
    "ID": "component",
    "routerData": Object {
      "activePageComponent": [Function],
      "activePageOptions": Object {
        "path": "base",
        "text": "Base page",
      },
      "url": "",
    },
  },
  "html": "
<h1>Title</h1>
<div>Content for dummy component</div>
",
}
`;
